packages/core-contracts/test/proxy/ForwardingProxy.test.js

```diff
@@ -0,0 +1,98 @@
+const { ethers } = hre;
+const assert = require('assert');
+const { assertRevert } = require('@synthetixio/core-js/utils/assertions');
+
+describe('ForwardingProxy', () => {
+  let ForwardingProxy, Instance, Implementation;
+
+  describe('when setting ImplementationMockA as the implementation', () => {
+    before('set up proxy and implementation', async () => {
+      let factory;
+
+      factory = await ethers.getContractFactory('ImplementationMockA');
+      Implementation = await factory.deploy();
+
+      factory = await ethers.getContractFactory('ForwardingProxyMock');
+      ForwardingProxy = await factory.deploy(Implementation.address);
+
+      Instance = await ethers.getContractAt('ImplementationMockA', ForwardingProxy.address);
+    });
+
+    it('shows that the implementation is set', async () => {
+      assert.equal(await ForwardingProxy.getImplementation(), Implementation.address);
+    });
+
+    describe('when interacting with the implementation via the proxy', async () => {
+      describe('when reading and setting a value that exists in the implementation', () => {
+        before('set a value', async () => {
+          await (await Instance.setA(42)).wait();
+        });
+
+        it('can read the value set', async () => {
+          assert.equal(await Instance.getA(), 42);
+        });
+      });
+
+      describe('when trying to call a function that is not payable', () => {
+        it('reverts', async () => {
+          await assertRevert(
+            Instance.setA(1337, { value: ethers.utils.parseEther('1') }),
+            'non-payable method'
+          );
+        });
+      });
+
+      describe('when reading and setting a value that does not exist in the implementation', () => {
+        let BadInstance;
+
+        before('wrap the implementation', async () => {
+          BadInstance = await ethers.getContractAt('ImplementationMockB', ForwardingProxy.address);
+        });
+
+        it('reverts', async () => {
+          await assertRevert(BadInstance.getB(), 'function selector was not recognized');
+        });
+      });
+    });
+  });
+
+  describe('when setting ImplementationMockB as the implementation', () => {
+    before('set up proxy and implementation', async () => {
+      let factory;
+
+      factory = await ethers.getContractFactory('ImplementationMockB');
+      Implementation = await factory.deploy();
+
+      factory = await ethers.getContractFactory('ForwardingProxyMock');
```

Comment:
if I understand correctly, here we don't want to create a new ForwardingProxy but use the one already created and just change the implementation.
ForwardingProxy.setImplementation(Implementation.address)

Reply:
No, `ForwardingProxy` is not intended to be upgradeable. The mock has that function, but the actual contract does and _should_ not.